InterviewPrep/chapter01/question0101.cs

```diff
@@ -1,26 +1,91 @@
-**/
-namespace InterviewPrep.chapter01
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright company="" file="Question0101.cs">
+//   
+// </copyright>
+// <summary>
+//   Question 0101: Does the string have unique characters?
+// </summary>
+// 
+// --------------------------------------------------------------------------------------------------------------------
+namespace InterviewPrep.Chapter01
 {
-    class question0101
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading.Tasks;
+
+    /// <summary> Question 0101: Does the string have unique characters? </summary>
+    public class Question0101
     {
+        /// <summary> Does the string have unique characters.    </summary>
+        /// <param name="s"> The s.   </param>
+        /// <exception cref="AggregateException"> The exception that contains all the individual exceptions thrown on all threads.  </exception>
+        /// <exception cref="OverflowException"> The dictionary already contains the maximum number of elements,     <see cref="F:System.Int32.MaxValue"/>. </exception> 
+        /// <exception cref="ArgumentNullException"> <paramref name="key"/> is a null reference (Nothing in Visual Basic).  </exception>
+        /// <returns> The <see cref="bool"/>.  </returns>
+        public static bool HasUniqueCharacters(string s)
+        {
+            // Contract.Requires(s != null);
+            if (CornerCase(s))
+            {
+                return true;
+            }
+
+            var cd = new ConcurrentDictionary<char, bool>();
+            var unique = true;
+            Parallel.ForEach(s, c =>
+                    {
+                        if (cd.ContainsKey(c))
+                        {
+                            unique = false;
 
-        static bool HasUniqueCharacters(String s)
+                            // TODO: Optimization: Add Cancellation token handling
+                        }
+
+                        cd.GetOrAdd(c, true);
+                    });
+            return unique;
+        }
+
+        /// <summary>
+        /// Does the string have unique characters
+        ///     What if you cannot use additional data structures?
+        /// </summary>
+        /// <param name="s"> The s.  </param>
+        /// <returns> The <see cref="bool"/>.  </returns>
+        public static bool HasUniqueCharactersSpaceEfficient(string s)
         {
+            // Contract.Requires(s != null);
+            if (CornerCase(s))
+            {
+                return true;
+            }
+
+            for (var i = 0; i < s.Length; i++)
+            {
+                for (var j = i + 1; j < s.Length; j++)
+                {
+                    if (s[i] == s[j])
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
```

Comment:
1. Assuming that the character set is fixed - can we do something better than N^2 solution . I am looking for an o(N) solution.